warp10/src/main/java/io/warp10/script/ext/token/TOKENDUMP.java

```diff
@@ -48,85 +48,93 @@
-  private byte[] tokenSipHashKey = null;
+  private final byte[] keystoreTokenAESKey;
+  private final byte[] keystoreTokenSipHashKey;
 
-  private boolean multikey = false;
+  /**
+   * Whether the keystore used to initialize the keys is that of a Warp/WarpDist instance.
+   * If true, a secret is needed to access the keystore keys.
+   */
+  private final boolean warpKeystore;
 
-  public TOKENDUMP(String name) {
+  public TOKENDUMP(String name, KeyStore keystore, boolean warpKeystore) {
     super(name);
-  }
-
-  public TOKENDUMP(String name, KeyStore keystore) {
-    super(name);
-    tokenAESKey = keystore.getKey(KeyStore.AES_TOKEN);
-    tokenSipHashKey = keystore.getKey(KeyStore.SIPHASH_TOKEN);
-  }
-
-  public TOKENDUMP(String name, boolean multikey) {
-    super(name);
-    this.multikey = multikey;
+    if(null != keystore) {
+      keystoreTokenAESKey = keystore.getKey(KeyStore.AES_TOKEN);
+      keystoreTokenSipHashKey = keystore.getKey(KeyStore.SIPHASH_TOKEN);
+    } else {
+      keystoreTokenAESKey = null;
+      keystoreTokenSipHashKey = null;
+    }
+    this.warpKeystore = warpKeystore;
   }
 
   @Override
   public Object apply(WarpScriptStack stack) throws WarpScriptException {
+    byte[] tokenAESKey = null;
+    byte[] tokenSipHashKey = null;
+    boolean customKeys = false;
 
-    byte[] AESKey = tokenAESKey;
-    byte[] SipHashKey = tokenSipHashKey;
-
-    if ((null == AESKey || null == SipHashKey) && !this.multikey) {
-      throw new WarpScriptException(getName() + " cannot be used in this context.");
-    }
+    // First, check if SipHash and AES keys are explicitly defined. In that case, no need for secret.
+    Object top = stack.pop();
+    if (top instanceof byte[]) {
+      customKeys = true;
 
-    Object top = null;
+      tokenSipHashKey = (byte[]) top;
 
-    if (this.multikey) {
       top = stack.pop();
 
       if (!(top instanceof byte[])) {
-        throw new WarpScriptException(getName() + " expects a SipHash Key (a byte array).");
+        throw new WarpScriptException(getName() + " expects a BYTES AES Key if a BYTES SipHash is given.");
       }
 
-      SipHashKey = (byte[]) top;
+      tokenAESKey = (byte[]) top;
 
       top = stack.pop();
+    }
 
-      if (!(top instanceof byte[])) {
-        throw new WarpScriptException(getName() + " expects an AES Key (byte array).");
-      }
-
-      AESKey = (byte[]) top;
-
-      top = stack.pop();
-    } else {
-      //
-      // A non null token secret was configured, check it
-      //
-      String secret = TokenWarpScriptExtension.TOKEN_SECRET;
+    if (null == tokenAESKey) { // in that case we have also null == tokenSipHashKey
+      // SipHash and AES keys are not explicitly defined, so we fall back to those of the keystore.
+      // Check the secret if needed before the fallback.
+      if (warpKeystore) {
+        String secret = TokenWarpScriptExtension.TOKEN_SECRET;
 
-      top = stack.pop();
+        if (null == secret) {
```

Comment:
You are implicitly making the definition of the secret mandatory, which was not the case previously, any reason for this change?

Reply:
This was already the case before but if was not clearly stated:

In TokenWarpScriptExtension.java when there was no secret in the conf nor OSS, the `keystore` was `null`:
`if (null == TOKEN_SECRET) { // if no configuration key` and `if (null != ossHandledSecret) {...} else { keystore = null; }`.

When the `keystore` was `null`, the `TokenWarpScriptExtension` constructor added the `TOKENGEN` and `TOKENDUMP` without keystore. It resulted in both keys to be `null` and `multikey` to be left to `false`. Then in `TOKENDUMP` we had
```
if ((null == AESKey || null == SipHashKey) && !this.multikey) {
      throw new WarpScriptException(getName() + " cannot be used in this context.");	
}
```
which resulted in an difficult-to-debug error message which in fact required the secret to be set.